### PR TITLE
Normalize count() variants

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Normalize count(constant), sum(1) to count(). This is needed for projection query routing.

Detailed description / Documentation draft:

.

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -377,6 +377,7 @@ class IColumn;
     M(Bool, optimize_respect_aliases, true, "If it is set to true, it will respect aliases in WHERE/GROUP BY/ORDER BY, that will help with partition pruning/secondary indexes/optimize_aggregation_in_order/optimize_read_in_order/optimize_trivial_count", 0) \
     M(UInt64, mutations_sync, 0, "Wait for synchronous execution of ALTER TABLE UPDATE/DELETE queries (mutations). 0 - execute asynchronously. 1 - wait current server. 2 - wait all replicas if they exist.", 0) \
     M(Bool, optimize_move_functions_out_of_any, false, "Move functions out of aggregate functions 'any', 'anyLast'.", 0) \
+    M(Bool, optimize_normalize_count_variants, false, "Rewrite aggregate functions that semantically equals to count() as count().", 0) \
```
Comment (vdimir) [inline src/Core/Settings.h:380]: Why we don't want to perform transformations unconditionally?

File: tests/queries/0_stateless/01706_optimize_normalize_count_variants.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+set optimize_normalize_count_variants = 1;
+
+explain syntax select count(), count(1), count(-1), sum(1);
```
Comment (vdimir) [inline tests/queries/0_stateless/01706_optimize_normalize_count_variants.sql:4]: Let's add count(null) to ensure that it is not optimized.

**Comment by vdimir** [comment]
Build error:
```
2021-02-07 08:25:47 ../src/Interpreters/RewriteCountVariantsVisitor.h:13:17: error: function 'DB::RewriteCountVariantsVisitor::visit' has a definition with different parameter names [readability-inconsistent-declaration-parameter-name,-warnings-as-errors]
2021-02-07 08:25:47     static void visit(ASTPtr & ast);
2021-02-07 08:25:47                 ^
2021-02-07 08:25:47 /build/obj-x86_64-linux-gnu/../src/Interpreters/RewriteCountVariantsVisitor.cpp:12:35: note: the definition seen here
2021-02-07 08:25:47 void RewriteCountVariantsVisitor::visit(ASTPtr & node)
2021-02-07 08:25:47                                   ^
2021-02-07 08:25:47 /build/obj-x86_64-linux-gnu/../src/Interpreters/RewriteCountVariantsVisitor.h:13:17: note: differing parameters are named here: ('ast'), in definition: ('node')
2021-02-07 08:25:47     static void visit(ASTPtr & ast);
```

File: src/Core/Settings.h (hunk 1 above)
Comment (amosbird) [reply src/Core/Settings.h:380]: good idea.

File: tests/queries/0_stateless/01706_optimize_normalize_count_variants.sql (hunk 2 above)
Comment (amosbird) [reply tests/queries/0_stateless/01706_optimize_normalize_count_variants.sql:4]: Absolutely.

**Review by vdimir — APPROVED** [review]

**Comment by amosbird** [comment]
Test failure looks unrelated.

**Comment by abyss7** [comment]
> Test failure looks unrelated.

It's related - it breaks master.

**Comment by amosbird** [comment]
> > Test failure looks unrelated.
> 
> It's related - it breaks master.

Oops. The test failure looks quite obscure w.r.t to parser changes. I still don't see how it's related.

**Comment by abyss7** [comment]
> > > Test failure looks unrelated.
> > 
> > 
> > It's related - it breaks master.
> 
> Oops. The test failure looks quite obscure w.r.t to parser changes. I still don't see how it's related.

With the new parser the `IFunction::arguments` may be `nullptr`

**Pull request merged**